### PR TITLE
Mccalluc/limit docker instances

### PR DESCRIPTION
**Pull request by mccalluc**
I think at this point the concern is only with imposing a limit on the container count. Now is not the time to figure out how errors from the API should be reported, and how the front end should handle them.

### CODE REVIEW

File: refinery/tool_manager/models.py [hunk 1]
```diff
@@ -413,6 +417,10 @@ def launch(self):
         launched container's url
             - <HttpResponseBadRequest>, <HttpServerError>
         """
+        client = DockerClientWrapper()
+        max = settings.DJANGO_DOCKER_ENGINE_MAX_CONTAINERS
```
Comment (scottx611x) [inline refinery/tool_manager/models.py]: could be confusing with Python's `max()` builtin?

File: refinery/tool_manager/tests.py [hunk 2]
```diff
@@ -2518,19 +2519,22 @@ def _start_visualization(
...
             self.assertEqual(post_response.status_code, 200)
 
-            tool = VisualizationTool.objects.get(
+            tools = VisualizationTool.objects.filter(
                 tool_definition__uuid=self.td.uuid
             )
-            self.assertEqual(tool.get_owner(), self.user)
+            if count:
+                self.assertEqual(len(tools), count)
+            last_tool = tools[len(tools)-1]
```
Comment (scottx611x) [inline refinery/tool_manager/tests.py]: Shorthand for this would be: `tools[:-1]`

File: refinery/tool_manager/models.py [hunk 3]
```diff
@@ -393,6 +393,10 @@ def update_file_relationships_with_urls(self):
         self.set_tool_launch_config(tool_launch_config)
 
 
+class VisualizationToolError(Exception):
```
Comment (scottx611x) [inline refinery/tool_manager/models.py]: If we are to have a custom exception, might want to inherit from a less-broad exception class

File: refinery/tool_manager/tests.py (hunk 2 above)
Comment (mccalluc) [reply refinery/tool_manager/tests.py]: not quite:
```
>>> a = [1,2,3]
>>> a[:-1]
[1, 2]
>>> a[-1:]
[3]
>>> a[-1:][0]
3
>>> a[-1]
3
```
But [alas](https://docs.djangoproject.com/en/1.11/topics/db/queries/#limiting-querysets):
> Negative indexing (i.e. Entry.objects.all()[-1]) is not supported

File: refinery/tool_manager/models.py (hunk 1 above)
Comment (mccalluc) [reply refinery/tool_manager/models.py]: good thought

File: refinery/tool_manager/tests.py (hunk 2 above)
Comment (hackdna) [reply refinery/tool_manager/tests.py]: [last()](https://docs.djangoproject.com/en/1.7/ref/models/querysets/#last) returns the last object in the queryset

Comment (scottx611x) [reply refinery/tool_manager/tests.py]: @mccalluc Oops, my mistake w/ the shorthand comment.

Regarding the `alas` though, one could do: `Entry.objects.last()`

**Comment by mccalluc** [comment]
Push failed with:
```
FAIL: test_invalid_TLC_non_LIST_PAIR_file_relationships (tool_manager.tests.ToolLaunchConfigurationTests)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/refinery-platform/refinery-platform/refinery/tool_manager/tests.py", line 2684, in test_invalid_TLC_non_LIST_PAIR_file_relationships

    context.exception.message

AssertionError: "The `file_relationships` defined didn't yield a valid LIST/PAIR nesting." not found in 'The \'file_relationships\' [{"Dicts aren\'t": \'LIST/PAIR-like\'}, {"Dicts aren\'t": \'LIST/PAIR-like\'}] type set([<type \'dict\'>]) is not a valid LIST/PAIR nesting: [set([<type \'list\'>]), set([<type \'tuple\'>])]'
```

Merge issue? Will look at right now.

**Pull request merged**